include/llbuild/CAS/CASDatabase.h

```diff
@@ -0,0 +1,70 @@
+//===- CASDatabase.h --------------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLBUILD_CAS_CASDATABASE_H
+#define LLBUILD_CAS_CASDATABASE_H
+
+#include "llbuild/CAS/DataID.h"
+
+#include "llbuild/Basic/Compiler.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/ErrorOr.h"
+
+#include <cstdint>
+#include <memory>
+#include <future>
+
+namespace llbuild {
+namespace CAS {
+
+struct CASObject;
+  
+/// An abstract content-addressable storage (CAS) database.
+class CASDatabase {
+private:
+  // Copying is disabled.
+  CASDatabase(const CASDatabase&) LLBUILD_DELETED_FUNCTION;
+  void operator=(const CASDatabase&) LLBUILD_DELETED_FUNCTION;
+
+public:
+  CASDatabase() {}
+  virtual ~CASDatabase();
+
+  /// Check if the database contains the given.
+  virtual auto contains(const DataID& id) ->
+    std::future<llvm::ErrorOr<bool>> = 0;
+
+  /// Get the given object, if present.
+  virtual auto get(const DataID& id) ->
+    std::future<llvm::ErrorOr<std::unique_ptr<CASObject>>> = 0;
+
+  /// Write the given object into the database.
+  virtual auto put(std::unique_ptr<CASObject> object) ->
```

Comment:
Reference?   That way the client doesn't have to relinquish its copy of the object on put.

Reply:
The big problem with that is that any implementation which wants to be async is going to have to copy the data, which might be a double copy if the client had to create an owning copy anyway?

This is part of what I meant with being a not very well though out API.

What I think we really want here is some kind of optionally-owning vector, like an ArrayRef but which can be a unique owner. I didn't want to write that, so I went with this for now.

Thoughts?